src/span.h

```diff
@@ -65,26 +131,20 @@ class Span
     template <typename O> friend class Span;
 };
 
-/** Create a span to a container exposing data() and size().
- *
- * This correctly deals with constness: the returned Span's element type will be
- * whatever data() returns a pointer to. If either the passed container is const,
- * or its element type is const, the resulting span will have a const element type.
- *
- * std::span will have a constructor that implements this functionality directly.
- */
-template<typename A, int N>
-constexpr Span<A> MakeSpan(A (&a)[N]) { return Span<A>(a, N); }
-
-template<typename V>
-constexpr Span<typename std::remove_pointer<decltype(std::declval<V>().data())>::type> MakeSpan(V& v) { return Span<typename std::remove_pointer<decltype(std::declval<V>().data())>::type>(v.data(), v.size()); }
+// MakeSpan helps constructing a Span of the right type automatically.
+/** MakeSpan for arrays: */
+template <typename A, int N> Span<A> constexpr MakeSpan(A (&a)[N]) { return Span<A>(a, N); }
+/** MakeSpan for temporaries / rvalue references, only supporting const output. */
+template <typename V> constexpr auto MakeSpan(V&& v) -> typename std::enable_if<!std::is_lvalue_reference<V>::value, Span<const typename std::remove_pointer<decltype(v.data())>::type>>::type { return std::forward<V>(v); }
+/** MakeSpan for (lvalue) references, supporting mutable output. */
+template <typename V> constexpr auto MakeSpan(V& v) -> Span<typename std::remove_pointer<decltype(v.data())>::type> { return v; }
```

Comment:
It's converting the argument to the return type, so this is effectively evaluating `Span<typename std::remove_pointer<decltype(v.data())>::type>(v)`. As `Span` has a constructor for converting from any data type that exposes a proper `size()` and `data()` member function (the range-like object constructor), so that's what is invoked here.

The reason for not calling the pointer/size constructor instead (which could have been written as `return {v.data(), v.size()};` btw), is that this would potentially bypass the protections that are already implemented for the range-like object constructor.

`MakeSpan`'s only purpose is automatic inference of a Span's type to construct, this way of implementing things guarantees it is not doing anything more than that (it doesn't change what can be constructed; it only lets you write `MakeSpan(...)` instead of `Span<...>(...)`).